### PR TITLE
[MI-2321] Added notification posts for pipeline subscription

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
Event types:
1. Build Completed
Screenshot from 2022-11-09 12-14-10

2. Release Created
Screenshot from 2022-11-09 12-16-20

3. Release Abandoned
Screenshot from 2022-11-09 12-17-04

4. Release Deployment Started
Screenshot from 2022-11-09 12-17-45

5. Release Deployment Completed
Screenshot from 2022-11-09 12-18-13

6. Run State Changed
Screenshot from 2022-11-09 12-14-51

7. Run Stage State Changed
Screenshot from 2022-11-09 12-15-27

### CODE REVIEW

File: server/plugin/api.go [hunk 1]
```diff
@@ -545,6 +564,205 @@ func (p *Plugin) handleSubscriptionNotifications(w http.ResponseWriter, r *http.
...
+					Title: "Requested for",
+					Value: body.Resource.RequestedFor.Name,
+					Short: true,
+				},
+				{
+					Title: "Duration",
+					Short: true,
+					Value: time.Time{}.Add(finishTime.Sub(startTime)).Format(constants.TimeLayout),
+				},
+			}, Footer: body.Resource.Project.Name,
```
Comment (ayusht2810) [inline server/plugin/api.go:608]:
```suggestion
			}, 
			Footer: body.Resource.Project.Name,
```

File: server/plugin/api.go [hunk 2]
```diff
@@ -545,6 +564,205 @@ func (p *Plugin) handleSubscriptionNotifications(w http.ResponseWriter, r *http.
...
+			FooterIcon: fmt.Sprintf(constants.StaticFiles, p.GetSiteURL(), constants.PluginID, constants.FileNameProjectIcon),
+		}
+	case constants.SubscriptionEventReleaseCreated:
+		artifacts := ""
+		for i := 0; i < len(body.Resource.Release.Artifacts); i++ {
+			if i != len(body.Resource.Release.Artifacts)-1 {
+				artifacts += fmt.Sprintf("%s, ", body.Resource.Release.Artifacts[i].Name)
+			} else {
+				artifacts += body.Resource.Release.Artifacts[i].Name
+			}
```
Comment (ayusht2810) [inline server/plugin/api.go:618]:
```suggestion
		for i := 0; i < len(body.Resource.Release.Artifacts); i++ {
		        artifacts += body.Resource.Release.Artifacts[i].Name
			if i != len(body.Resource.Release.Artifacts)-1 {
				artifacts += ", "
			}
```

File: server/serializers/subscriptions.go [hunk 3]
```diff
@@ -107,10 +107,55 @@ type Resource struct {
...
+	Reason            string      `json:"reason"`
+	ModifiedOn        string      `json:"modifiedOn"`
+	ModifiedBy        Reviewer    `json:"modifiedBy"`
+	Links             ProjectLink `json:"_links"`
+}
+
+type Artifact struct {
+	Name string `json:"alias"`
+}
+type RequestedFor struct {
```
Comment (ayusht2810) [inline server/serializers/subscriptions.go:151]:
```suggestion
}

type RequestedFor struct {
```

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/api.go [hunk 4]
```diff
@@ -496,6 +499,22 @@ func (p *Plugin) handleSubscriptionNotifications(w http.ResponseWriter, r *http.
 			sourceBranchName = strings.Split(body.Resource.PullRequest.SourceRefName, "/")[2]
 		}
 
+		// Convert map to json string
+		jsonStr, err := json.Marshal(body.Resource.Comment)
```
Comment (Kshitij-Katiyar) [inline server/plugin/api.go:503]: i think json.Marshal returns a byte array and not a string so this variable name could be misleading

File: server/plugin/api.go [hunk 5]
```diff
@@ -545,6 +564,205 @@ func (p *Plugin) handleSubscriptionNotifications(w http.ResponseWriter, r *http.
 			Footer:     fmt.Sprintf("%s | %s", strings.Split(body.Resource.RefUpdates[0].Name, "/")[2], body.Resource.Repository.Name),
 			FooterIcon: fmt.Sprintf(constants.StaticFiles, p.GetSiteURL(), constants.PluginID, constants.FileNameGitBranchIcon),
 		}
+	case constants.SubscriptionEventBuildCompleted:
+		startTime, err := time.Parse(constants.DateTimeLayout, strings.Split(body.Resource.StartTime, ".")[0])
```
Comment (Kshitij-Katiyar) [inline server/plugin/api.go:568]: are we checking that the the array should have atlease one element to access

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/plugin/api.go (hunk 5 above)
Comment (raghavaggarwal2308) [reply server/plugin/api.go:568]: `StartTime` and `FinishTime` will never be empty. So, it will always contain at least one value

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/api.go [hunk 6]
```diff
@@ -545,6 +564,205 @@ func (p *Plugin) handleSubscriptionNotifications(w http.ResponseWriter, r *http.
...
+				},
+				{
+					Title: "Requested for",
+					Value: body.Resource.RequestedFor.Name,
+					Short: true,
+				},
+				{
+					Title: "Duration",
+					Short: true,
+					Value: time.Time{}.Add(finishTime.Sub(startTime)).Format(constants.TimeLayout),
```
Comment (avas27JTG) [inline server/plugin/api.go:606]:
```suggestion
					Value: time.Time{}.Add(finishTime.Sub(startTime)).Format(constants.TimeLayout),
				        Short: true,
```
for consistency

File: server/serializers/subscriptions.go [hunk 7]
```diff
@@ -107,10 +107,56 @@ type Resource struct {
...
+
+type Environment struct {
+	Release           Release    `json:"release"`
+	ReleaseDefinition Definition `json:"releaseDefinition"`
+}
+
+type Release struct {
+	Name              string      `json:"name"`
+	CreatedBy         Reviewer    `json:"createdBy"`
+	Artifacts         []Artifact  `json:"artifacts"`
```
Comment (avas27JTG) [inline server/serializers/subscriptions.go:140]:
```suggestion
	Artifacts         []*Artifact  `json:"artifacts"`
```

**Review by Kshitij-Katiyar — APPROVED** [review]

**Review by avas27JTG — APPROVED** [review]

**Pull request merged**